### DISCUSSION TITLE
How to improve API response time for AI-powered web application with OpenAI GPT Chat Completion?

### DISCUSSION THREAD

**Discussion by yaschet** [community/community#64299] [Programming Help] [answered]
### **Preface:**
We have developed an AI-powered web application with the goal of processing scanned student certificates and grade transcripts. These documents are uploaded to a NoSQL database (specifically MongoDB). We use OCR (Optical Character Recognition) for data extraction, then employ OpenAI LLM Models to convert the extracted information to JSON. Additionally, we leverage AI for translating the JSON content into English.

### **Problem:**
During the use of OpenAI GPT Chat Completion, there are occasional delays in response time, sometimes exceeding 30 seconds. This occurs approximately 70% of the time during the translation phase of the JSON object and 30% during the conversion phase from a string to JSON.

We analyzed and adopted the [gpt-3.5-turbo-0301] model to mitigate this issue due to its faster response time while yielding the exact expected output.

Our Python Flask Application features 8 API endpoints:
- Initialization: "/api/v1/initialize" [0.5-3 seconds].
- Text Extraction: "/api/v1/extract-text" [4-15 seconds].
- Table Extraction: "/api/v1/extract-tables" [8-20 seconds].
- Text Correction: "/api/v1/correct-text" [2-20 seconds].
- Table Correction: "/api/v1/correct-tables" [13-30+ seconds].
- Text Translation: "/api/v1/translate-text" [2-20 seconds].
- Table Translation: "/api/v1/translate-tables" [13-30+ seconds].
- Document Generation: "/api/v1/generate-document" [5-10 seconds].

Currently, the issue is concentrated on two endpoints: "/api/v1/correct-tables" and "/api/v1/translate-tables".

As our client side is developed with Next.js, communication with the server is solely through HTTP. It's important to note that we are constrained by Heroku's requested timeout limit of 30 seconds.

### **Proposed solutions:**
1. Enhance the client-side logic to handle process failures without triggering timeout errors. Failed processes would initiate a 5-second waiting period followed by automatic reattempting until success.
2. Explore the possibility of switching to a different GPT provider due to perceived intentional slowdowns in OpenAI's API responses. However, this alternative may not guarantee a solution and require further evaluation.

Despite attempting multiple solutions, the core issue remains the duration of API responses. We seek insights from experienced professionals to identify more effective workarounds for this challenge.

I sincerely appreciate your time and eagerly await your valuable advice!

**Comment by viraj-s15** [comment, marked as answer]
I ran into similar issue when using a self hosted RAG using llama2. I ended up removing all unnecessary props in the JSON and removing all newlines/tabs in the data. I also made the JSON props less verbose. For example, the prop "response" became "res","value" became "v" and so on. You may also want to reduce the length of response itself by using stop sequences. You could also implement a middleware that checks for the response periodically.
